lib/Transforms/SimplifySelect.cpp

```diff
@@ -0,0 +1,686 @@
+
+        return mlir::success();
+    }
+
+ private:
+    std::tuple<mlir::Value, mlir::Value, bool> createMaskOp(mlir::PatternRewriter &rewriter,
+                                                            P4HIR::BitsType destType,
+                                                            mlir::Value yieldArg) const {
+        mlir::Value keyVal;
+        mlir::Value keyMask;
+        bool needsMask = false;
+
+        if (auto constOp = yieldArg.getDefiningOp<P4HIR::ConstOp>()) {
+            llvm::APInt val;
+            llvm::APInt mask;
+
+            if (P4HIR::isUniversalSetValue(yieldArg)) {
+                val = mask = llvm::APInt::getZero(destType.getWidth());
+                needsMask = true;
+            } else {
+                auto setAttr = constOp.getValueAs<P4HIR::SetAttr>();
+
+                if (setAttr.getKind() == P4HIR::SetKind::Constant) {
+                    val = setAttr.getConstantValue<P4HIR::IntAttr>();
+                    mask = llvm::APInt::getAllOnes(val.getBitWidth());
+                } else {
+                    std::tie(val, mask) = setAttr.getMaskValues();
+                    needsMask = true;
+                }
+            }
+
+            keyVal = rewriter.create<P4HIR::ConstOp>(yieldArg.getLoc(),
+                                                     P4HIR::IntAttr::get(destType, val));
+            keyMask = rewriter.create<P4HIR::ConstOp>(yieldArg.getLoc(),
+                                                      P4HIR::IntAttr::get(destType, mask));
+        } else if (auto maskOp = yieldArg.getDefiningOp<P4HIR::MaskOp>()) {
+            keyVal = maskOp.getLhs();
+            keyMask = maskOp.getRhs();
+            needsMask = true;
+        } else if (auto setOp = yieldArg.getDefiningOp<P4HIR::SetOp>()) {
+            auto mask = llvm::APInt::getAllOnes(destType.getWidth());
+            keyVal = setOp.getInput()[0];
+            keyMask = rewriter.create<P4HIR::ConstOp>(yieldArg.getLoc(),
+                                                      P4HIR::IntAttr::get(destType, mask));
+        } else {
+            llvm_unreachable("Impossible key value.");
+        }
+
+        return {keyVal, keyMask, needsMask};
+    }
+};
+
+}  // end namespace
+
+void SimplifySelect::runOnOperation() {
+    mlir::RewritePatternSet patterns(&getContext());
+
+    if (flattenTuples) patterns.add<FlattenTuples>(patterns.getContext());
+    if (replaceBools) patterns.add<ReplaceBoolWithInt>(patterns.getContext());
+    if (replaceRanges)
+        patterns.add<ReplaceRanges>(patterns.getContext(), replaceRangesCaseCountLimit);
+    if (concatArgs) patterns.add<MakeSingleArgument>(patterns.getContext());
+
+    // Collect operations and apply patterns.
+    llvm::SmallVector<mlir::Operation *, 16> ops;
+    getOperation()->walk([&](P4HIR::ParserTransitionSelectOp op) { ops.push_back(op); });
+
+    mlir::GreedyRewriteConfig config;
+    config.strictMode = mlir::GreedyRewriteStrictness::ExistingOps;
```

Comment:
What if we'd have bool inside tuple as a key? Won't this prevent application of second pattern to the newly created op? Should we use `ExistingAndNewOps` here?